traffic-mirroring/mirroring_egress_kern.c

```diff
@@ -0,0 +1,269 @@
+ *  $TC filter  del dev $DEV egress
+ *
+ * Does TC redirect respect IP-forward settings?
+ *
+ */
+static __always_inline int egress_redirect(struct __sk_buff *skb)
+{
+    bool src_addr_match = false;
+    bool proto_match = false;
+    bool src_port_match = false;
+    bool dst_port_match = false;
+
+    bool allow_all_ip = false;
+    bool allow_all_src_ports = false;
+    bool allow_all_dst_ports = false;
+
+    int *ifindex;
+    int *ifany;
+    int iface_key = 1;
+
+    /* Lookup what ifindex to redirect packets to */
+    ifindex = bpf_map_lookup_elem(&redirect_iface, &iface_key);
+    if (!ifindex) {
+        return TC_ACT_OK;
+    }
+
+    void *data = (void *)(long)skb->data;
+    void *data_end = (void *)(long)skb->data_end;
+
+    const int l3_off = ETH_HLEN; // IP header offset
+    const int l4_off = l3_off + sizeof(struct iphdr); // TCP header offset
+
+    struct ethhdr *eth = data;
+    int l7_off;
+
+    if (data + sizeof(*eth) > data_end) {
+        return TC_ACT_OK;
+    }
+    if (eth->h_proto != htons(ETH_P_IP)) {
+        return TC_ACT_OK; // Not an IPv4 packet, handover to kernel
+    } else if (eth->h_proto == htons(ETH_P_ARP)) {
+        return TC_ACT_OK;
+    }
+
+    if (data + l4_off > data_end) {
+        return TC_ACT_OK; // Not our packet, handover to kernel
+    }
+    struct iphdr *iph = (struct iphdr *)(data + l3_off);
+
+    if (iph->protocol == IPPROTO_TCP) {
+        l7_off = l4_off + sizeof(struct tcphdr); // L7 (e.g. HTTP) header offset
+    } else if (iph->protocol == IPPROTO_UDP) {
+        l7_off = l4_off + sizeof(struct udphdr); // L7 (e.g. HTTP) header offset
+    } else if (iph->protocol == IPPROTO_ICMP) {
+        l7_off =
+            l4_off + sizeof(struct icmphdr); // L7 (e.g. HTTP) header offset
+    } else {
+        return TC_ACT_OK;
+    }
+
+    if (data + l7_off > data_end) {
+        return TC_ACT_OK; // Not our packet, handover to kernel
+    }
+
+    uint32_t all_ports = 0;
+    uint32_t fin_srcport = 0;
+    uint32_t fin_dstport = 0;
+
+    __u32 daddr = iph->daddr;
+
+    struct daddr_key dkey;
+    dkey.prefix_len = 32;
+    dkey.data[0] = daddr & 0xff;
+    dkey.data[1] = (daddr >> 8) & 0xff;
+    dkey.data[2] = (daddr >> 16) & 0xff;
+    dkey.data[3] = (daddr >> 24) & 0xff;
+
+    // Binary representation of ifany variable. Reading from LSB to MSB:
+    //      0th bit True: allow all/any IP addresses
+    //      1th bit True: allow all/any source port
+    //      2th bit True: allow all/any destination port
+    ifany = bpf_map_lookup_elem(&egress_any, &iface_key);
+    if (ifany) {
+        if (((*ifany) >> (0)) % 2 == 1) {
+            allow_all_ip = true;
+        }
+        if (((*ifany) >> (1)) % 2 == 1) {
+            allow_all_src_ports = true;
+        }
+        if (((*ifany) >> (2)) % 2 == 1) {
+            allow_all_dst_ports = true;
+        }
+    }
+
+    /* Check if the packet's destination ip falls in the remote endpoint CIDR */
```

Comment:
```suggestion
    /* Check if the packet's destination IP address falls in the remote endpoint prefix */
```
Use IP version agnostic terminology where possible